torch/csrc/jit/passes/quantization.cpp

```diff
@@ -87,8 +164,89 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
   }
 }
 
-void InsertFakeQuantNodes(std::shared_ptr<Graph>& graph) {
-  throw std::runtime_error("Pass not implemented yet!");
+void InsertQuantDequantNodes(std::shared_ptr<Graph>& graph) {
+  std::stack<Block*> blocks_to_visit;
+  blocks_to_visit.push(graph->block());
+  // For storing quantizable values - node pairs that are external
+  // or intermediate inputs to quantizable nodes
+  std::vector<std::pair<Value*, Node*>> quantInputs;
+  // For storing quantizable values that are output of quantizable nodes
+  std::vector<Value*> quantOutputs;
+
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+
+    // Block external inputs denoted as prim::Param node do not
+    // belong to Block list. We iterate over the inputs Values and
+    // its uses. If the user is a quantizable node, we cache it to
+    // quantizableInput vector as a unique V,N pair
+    // Example : N1(prim::Param) -> (V1 -> (N4, N5), V2 -> (N6, N7), V3)
+    //           N1 is not quantizable node but N4 and N7 are
+    //           quantizable nodes. So we cache the (V1, N4) and
+    //           (V2, N7) as insertion branch for quant-dequant nodes
+    auto inputVals = b->inputs();
+    for (auto& v : inputVals) {
+      if (checkifExternalInput(v) &&
+        v->type()->isSubtypeOf(TensorType::get())) {
+        for (auto& use : v->uses()) {
+          if (checkIfNodeQuantizable(use.user)) {
+            quantInputs.emplace_back(std::make_pair(v, use.user));
+          }
+        } // end inner for
+      }
+    } //end outer for
+
+    for (Node* n : b->nodes()) {
+      // Schedule the sub blocks
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
+
+      // Next we iterate over the outputs of quantizable node
+      // and cache it to the quantizableOutput vector.
+      // Example N1 -> (V1 -> (N4, N5), V2 -> (N6, N7), V3)
+      //         N1 is quantizable node. So we insert quant-dequant
+      //         nodes for all outputs of N1
+      if (checkIfNodeQuantizable(n)) {
+        for (auto& v : n->outputs()) {
+          // Quantizer will look at all Tensor subTypes of outputs
+          // for quantization
+          if (v->type()->isSubtypeOf(TensorType::get())) {
+            quantOutputs.emplace_back(v);
+          }
+        }
+      } else {
+        // This iteration is similar to external inputs to block.
+        // For non quantizable node within block, we iterate over
+        // outputs and if uses are inputs to quantizable node, we cache it to
+        // quantizableInput vector as a unique V,N pair
+        // Example N1(Belong to Block) -> (V1 -> (N4, N5), V2 -> (N6, N7), V3)
+        //         N1 is not quantizable node. N4 and N7 are quantizable nodes.
+        //         So we cache the (V1, N4) and (V2, N7) as insertion branch
+        //         for quant-dequant nodes
+        for (auto& v : n->outputs()) {
```

Comment:
Please use `Value *` instead of `auto&`, it's easier to read (here and in other similar places).